docs/proposals/checks_as_wheels.md

```diff
@@ -0,0 +1,288 @@
+dependencies but we implement the logic elsewhere.
+
+### User experience: final user
+
+At the moment we are exposed to weird corner cases when it comes to the point of
+installing checks in standalone mode, let’s see an example:
+
+ * User installs agent 5.0.0 shipping ntp check 1.0.0
+ * Agent 5.1.0 gets released, shipping ntp check 1.1.0
+ * User updates **only** the ntp check to 1.1.0
+ * Agent 5.2.0 gets released, shipping ntp check 1.2.0
+ * User updates the agent to 5.2.0 but still runs ntp 1.1.0 (standalone checks
+ take precedence)
+
+We have strategies in place to mitigate the effect of the example above, but the
+problem still exists.
+
+### User experience: developer
+
+Most of the checks are supposed to run both unit and integration tests, at the
+moment they are merged in a single test file run by our CI. A contributor needs
+to use our ruby-based tooling in order to run tests, even in the case they only
+want to run unit tests.
+
+Most if not all the checks are not supposed to run outside the agent lifecycle,
+working on a brand new check or trying to change an existing one heavily rely on
+this, making things more complicated than they could be.
+
+A couple of related, minor issues it would be nice to fix: at the moment it’s not
+possible to split the code of a check across multiple Python modules, same
+happens for the tests.
+
+## Constraints
+
+Any solution must work with the new agent, working for both the old and the new
+one is preferable.
+
+## Recommended Solution
+
+### Package checks as Python Wheels
+
+The solution proposed is to migrate packaging and distribution of our checks
+from system packages built with Omnibus to _universal Python Wheels_ when possible,
+and _binary Wheels_ when we need to ship non-python artifacts.
+
+The changes needed in the agent to support this additional method of checks
+distribution are the following:
+https://github.com/DataDog/dd-agent/compare/massi/whl-poc
+The proof of concept should be extended to rework a bit the loading logic but
+it already works.
+
+The work needed to migrate a pure Python check to wheels is not that much, see
+the ntp check for an example:
+https://github.com/DataDog/integrations-core/tree/massi/whl-poc/ntp
+
+Moving to wheels wouldn’t solve any possible problem we’re facing in building
+and shipping `integrations-core` and `integrations-extra` but overall it might
+work better, let’s see few examples.
+
+#### Building
+
+Building a pure Python check would be as easy as doing
+`python setup.py bdist_wheel`. In the case of checks that ship binary artifacts
+more work would be needed to set up the build scripts, still the build strategy
+would be the same. A locally built check is basically a zip archive, so it could
+be easily moved around and picked up by an agent. In the case of a pure python
+check, the archives would be platform independent, meaning one could build the
+check on a Mac and copy and install the artifact in a Linux box.
+
+#### Versioning
```

Comment:
This is very nice when we look at both `dd-agent` and `integrations-core` it's very difficult to know which version was shipped (even more when looking only at the history of a specific integration).